File: euler/94.c
#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>
#include <math.h>

uint64_t isqrt(uint64_t a, bool* isinteger)
{
    uint64_t x = sqrt(a);

    *isinteger = x*x == a;

    return x;
}


int main(void)
{
    uint64_t sum = 0;

    for(uint64_t a = 1; true; a++)
    {
        uint64_t array[2] = { a-1, a+1 };

        for(int i = 0; i < 2; i++)
        {
            uint64_t c = array[i];
            bool isinteger;

            uint64_t value = isqrt(4*a*a-c*c, &isinteger);

            // A * 4
            uint64_t A4 = value*c;

            uint64_t U = a+a+c;
            if(U >= 1000000000)
                goto out;

            if(isinteger && A4 && (A4 % 4) == 0)
                sum += U;
        }
    }

out:
    printf("%llu\n", (long long unsigned)sum);

    return 0;
}

File: euler/94.py
import math

def sqrt(a):
    x = int(round(math.sqrt(a)))

    if x*x == a:
        return True, x
    else:
        return False, x


if __name__ == "__main__":
    s = 0
    a = -1

    print("# a, c, A, U")
    while True:
        a += 2

        for c in (a-1, a+1):
            isinteger,value = sqrt(4*a*a-c*c)
            #print(isinteger, 4*a*a-c*c, math.sqrt(4*a*a-c*c))

            # A * 4
            A4 = value*c

            U = a+a+c
            if U >= 1000000000:
                break

            if isinteger and A4 and A4 % 4 == 0:
                A = A4//4
                s += U
                print(a, c, A, U)
                #print(a, c, sqrt(a**2-(c/2)**2))

    print(s)
